File: algorithms/dynamicProgramming/UnboundedKnapsack.java
import java.io.*;
import java.util.*;
import java.text.*;
import java.math.*;
import java.util.regex.*;

// Solution to https://www.hackerrank.com/challenges/unbounded-knapsack

public class Solution {

    public static void main(String[] args) {
        Scanner in = new Scanner(System.in);
        int T = in.nextInt();
        for(int i = 0; i < T; i++){
            int N = in.nextInt();
            int K = in.nextInt();
            int[] arr = new int[N];
            for(int j = 0; j < N; j++){
                arr[j] = in.nextInt();
            }

            boolean[] ans = new boolean[K+1];
            ans[0] = true;
            int max = 0;
            for(int j = 1; j <= K; j++){
                for(int k = 0; k < N; k++){
                    if(j-arr[k] >= 0 && ans[j-arr[k]]){
                        ans[j] = true;
                        max = j;
                        break;
                    }
                }
            }
            System.out.println(max);
        }
    }
}

File: algorithms/dynamicProgramming/unbounded_knapsack.py
# Solution to https://www.hackerrank.com/challenges/unbounded-knapsack

t = int(input())
for _ in range(t):
    n,k = [int(x) for x in input().split()]
    a = [int(x) for x in input().split()]

    ans = 0
    # cache[i]: whether i can be made using elements in A
    cache = [False for x in range(k+1)]
    # base case: 0 can be made by selecting no items in A
    cache[0] = True
    # Induction: i can be made if an element in A, j, exists such that
    # i - j >= 0 and i-j can be made
    for i in range(1, k+1):
        for j in a:
            if i - j >= 0 and cache[i-j]:
                cache[i] = True
                ans = i
    print(ans)
